File: crates/compiler/frontend/python/apxm_program/_bound_tree.py
"""Immutable typed source tree for authored Agent programs.

The tree is the frontend's semantic representation between the native Python AST
and the language-neutral FrontendGraph. It retains lexical constructs, resolved
declarations, inferred type references, and source spans. It carries no AIR or
AIS operation name, no runtime value, and no mutable recorder state; it is frozen
after construction so traversal cannot depend on marker execution order.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Optional, Union

from ._generated.frontend_records import CallIntent, ControlIntent
from ._generated.permissions import Permission


@dataclass(frozen=True, slots=True)
class Span:
    """A source location range for one construct."""

    source_file: str
    start_line: int
    start_column: int
    end_line: int
    end_column: int


@dataclass(frozen=True, slots=True)
class BoundParameter:
    """One typed callback parameter with its resolved role."""

    value_id: str
    type_ref: str
    role: str  # agent_facade | input | context | ordinary


@dataclass(frozen=True, slots=True)
class BoundDeclaration:
    """A resolved Context, Model, Tool, Capability, or Event declaration."""

    decl_id: str
    decl_kind: str  # context | model_binding | tool_binding | capability_binding | event_type
    input_type_ref: str
    output_type_ref: str
    target_ref: Optional[str] = None
    context_default_present: Optional[bool] = None


@dataclass(frozen=True, slots=True)
class BoundValue:
    """One typed value with exactly one origin."""

    value_id: str
    type_ref: str
    origin: str  # parameter | call_result | block_argument | context_value | literal | resume_input
    origin_id: Optional[str] = None
    expression: Optional[dict[str, object]] = None


@dataclass(frozen=True, slots=True)
class BoundOperand:
    """A typed use of a value at a named consumer slot."""

    value_id: str
    slot: str


@dataclass(frozen=True, slots=True)
class BoundPredicate:
    """A closed structural predicate over one prior typed value."""

    root_value_id: str
    property_path: tuple[str, ...]
    comparator: str  # truthy | equals | not_equals
    literal: Optional[dict[str, object]] = None


@dataclass(frozen=True, slots=True)
class BoundCall:
    """A resolved effectful call: model, tool, capability, agent, or event.

    Carries a generated ``CallIntent`` contract record plus the authoring-time
    extras the contract does not state: a source ``span`` and typed
    ``operands`` with consumer-slot identity (the contract instead carries
    ``operand_values`` and leaves slot identity to ``data_edges``). The flat
    properties below read through to ``contract`` so every other reader of a
    ``BoundCall`` is unaffected by this split.
    """

    contract: CallIntent
    span: Optional[Span]
    operands: tuple[BoundOperand, ...]

    @property
    def node_id(self) -> str:
        return self.contract.node_id

    @property
    def intent_kind(self) -> str:
        return self.contract.intent_kind

    @property
    def parent_region_id(self) -> str:
        return self.contract.parent_region_id

    @property
    def execution_order(self) -> int:
        return self.contract.execution_order

    @property
    def binding_ref(self) -> Optional[str]:
        return self.contract.binding_ref

    @property
    def receiver_kind(self) -> Optional[str]:
        return self.contract.receiver_kind

    @property
    def result_value(self) -> Optional[str]:
        return self.contract.result_value


@dataclass(frozen=True, slots=True)
class BoundControl:
    """A resolved structural construct: branch, loop, task group, try, yield, return.

    Carries a generated ``ControlIntent`` contract record plus the
    authoring-time extras the contract does not state: a source ``span`` and
    typed ``operands`` with consumer-slot identity. ``predicate`` is kept as
    its own field rather than populated on ``contract``: the generated
    ``ControlIntent.predicate`` is typed as the generated ``ControlPredicate``
    union, whose ``EqualsPredicate``/``NotEqualsPredicate`` branches carry a
    generated ``PredicateLiteral``, and threading capture's predicate
    construction through those generated types is out of this change's scope
    (see the frontend-vocabulary-generation design note, §4 item 2). The flat
    properties below read through to ``contract`` so every other reader of a
    ``BoundControl`` is unaffected by this split.
    """

    contract: ControlIntent
    span: Optional[Span]
    operands: tuple[BoundOperand, ...]
    predicate: Optional[BoundPredicate]

    @property
    def node_id(self) -> str:
        return self.contract.node_id

    @property
    def control_kind(self) -> str:
        return self.contract.control_kind

    @property
    def parent_region_id(self) -> str:
        return self.contract.parent_region_id

    @property
    def execution_order(self) -> int:
        return self.contract.execution_order

    @property
    def body_region_ids(self) -> tuple[str, ...]:
        return self.contract.body_region_ids or ()

    @property
    def result_value(self) -> Optional[str]:
        return self.contract.result_value


@dataclass(frozen=True, slots=True)
class BoundRegion:
    """One lexical region owning ordered children."""

    region_id: str
    region_role: str  # one of RegionRole in _generated.frontend_graph
    parent_region_id: Optional[str]
    execution_order: int


@dataclass(frozen=True, slots=True)
class BoundHook:
    """A static before/after Hook binding and its captured body region."""

    hook_id: str
    scope: str
    phase: str
    target_selector: str
    declaration_order: int
    handler_ref: str
    handler_digest: str
    input_type_ref: str
    output_type_ref: str
    return_mode: str
    body_region_id: str
    assigned_context_value_id: Optional[str] = None


@dataclass(frozen=True, slots=True)
class BoundContextEdge:
    """An explicit typed Context transition between two nodes."""

    from_node: str
    to_node: str
    context_type_ref: str
    value_id: str


@dataclass(frozen=True, slots=True)
class BoundCapabilityRequirement:
    """One authored Capability declaration.

    Declarations are held one per authored binding, never one per
    ``capability_ref``: the same capability declared as both a Tool and a plain
    Capability is two distinct requirements and both reach the FrontendGraph.
    """

    capability_ref: str
    tool_schema_present: bool
    requested_permission: Optional[Permission] = None


@dataclass(frozen=True, slots=True)
class BoundProgram:
    """The immutable bound tree for one authored Agent program."""

    program_id: str
    entrypoint: str
    input_type_ref: str
    output_type_ref: str
    has_default_context: bool
    context_type_ref: Optional[str]
    parameters: tuple[BoundParameter, ...]
    body_region_id: str
    declarations: tuple[BoundDeclaration, ...] = ()
    values: tuple[BoundValue, ...] = ()
    regions: tuple[BoundRegion, ...] = ()
    calls: tuple[BoundCall, ...] = ()
    controls: tuple[BoundControl, ...] = ()
    context_edges: tuple[BoundContextEdge, ...] = ()
    hooks: tuple[BoundHook, ...] = ()
    imported_programs: tuple[tuple[str, str, str, str], ...] = ()
    capability_requirements: tuple[BoundCapabilityRequirement, ...] = ()
    model_requirements: tuple[str, ...] = ()
    spans: tuple[tuple[str, Span, str], ...] = field(default=())


BoundNode = Union[BoundCall, BoundControl]

File: crates/compiler/frontend/typescript/src/bound-tree.ts
// Immutable typed source tree for authored Agent programs.
//
// Mirrors `_bound_tree.py`: the frontend's semantic representation between
// the parsed AST and the language-neutral FrontendGraph. It retains lexical
// constructs, resolved declarations, inferred type references, and source
// spans. It carries no AIR or AIS operation name, no runtime value, and no
// mutable recorder state.
//
// `capture.ts` builds this tree and `emit.ts` folds it, exactly as
// `_capture.py` and `_emit.py` do, which is what makes the two frontends
// structurally the same below the AST walk rather than only reviewed by
// inspection.
//
// Where `_bound_tree.py` documents a closed set in a trailing comment, this
// file binds the generated vocabulary type instead — TypeScript can state it
// and Python cannot until the capture threads its records (design note, §4
// item 2), so the difference is one of what each language can express, not of
// what either tree holds.

import type {
  DeclKind,
  HookPhase,
  HookReturnMode,
  HookScope,
  InputContract,
  ParameterRole,
  PredicateComparator,
  RegionRole,
  ValueOrigin,
} from "./generated/frontend-graph.js";
import type {
  CallIntent,
  ControlIntent,
  PredicateLiteral,
  SkillRequirement,
  ValueExpression,
} from "./generated/frontend-records.js";
import type { Permission } from "./generated/permissions.js";

/** A source location range for one construct. */
export type Span = {
  readonly source_file: string;
  readonly start_line: number;
  readonly start_column: number;
  readonly end_line: number;
  readonly end_column: number;
};

/** One typed callback parameter with its resolved role. */
export type BoundParameter = {
  readonly value_id: string;
  readonly type_ref: string;
  readonly role: ParameterRole;
};

/** A resolved Context, Model, Tool, Capability, or Event declaration. */
export type BoundDeclaration = {
  readonly decl_id: string;
  readonly decl_kind: DeclKind;
  readonly input_type_ref: string;
  readonly output_type_ref: string;
  readonly target_ref?: string;
  readonly context_default_present?: boolean;
};

/** One typed value with exactly one origin. */
export type BoundValue = {
  readonly value_id: string;
  readonly type_ref: string;
  readonly origin: ValueOrigin;
  readonly origin_id?: string;
  readonly expression?: ValueExpression;
};

/** A typed use of a value at a named consumer slot. */
export type BoundOperand = {
  readonly value_id: string;
  readonly slot: string;
};

/**
 * A closed structural predicate over one prior typed value.
 *
 * The comparator is carried dynamically rather than as one of the contract's
 * three `ControlPredicate` branches, mirroring `_bound_tree.BoundPredicate`:
 * capture decides the branch from the authored comparison operator, and the
 * generated union serializer dispatches on the same discriminant when the
 * predicate reaches the graph.
 */
export type BoundPredicate = {
  readonly root_value_id: string;
  readonly property_path: readonly string[];
  readonly comparator: PredicateComparator;
  readonly literal?: PredicateLiteral;
};

/**
 * A resolved effectful call: model, tool, capability, agent, or event.
 *
 * Carries a generated `CallIntent` contract record plus the authoring-time
 * extras the contract does not state: a source `span` and typed `operands`
 * with consumer-slot identity (the contract instead carries `operand_values`
 * and leaves slot identity to `data_edges`). Mirrors `_bound_tree.BoundCall`.
 */
export type BoundCall = {
  readonly contract: CallIntent;
  readonly span?: Span;
  readonly operands: readonly BoundOperand[];
};

/**
 * A resolved structural construct: branch, loop, task group, try, yield, return.
 *
 * Carries a generated `ControlIntent` contract record plus the
 * authoring-time extras the contract does not state: a source `span` and
 * typed `operands` with consumer-slot identity. `predicate` is kept as its
 * own field rather than populated on `contract`, mirroring
 * `_bound_tree.BoundControl`; `emit.ts` puts it back on the contract record
 * when it folds the tree.
 */
export type BoundControl = {
  readonly contract: ControlIntent;
  readonly span?: Span;
  readonly operands: readonly BoundOperand[];
  readonly predicate?: BoundPredicate;
};

/** One lexical region owning ordered children. */
export type BoundRegion = {
  readonly region_id: string;
  readonly region_role: RegionRole;
  readonly parent_region_id?: string;
  readonly execution_order: number;
};

/** A static before/after Hook binding and its captured body region. */
export type BoundHook = {
  readonly hook_id: string;
  readonly scope: HookScope;
  readonly phase: HookPhase;
  readonly target_selector: string;
  readonly declaration_order: number;
  readonly handler_ref: string;
  readonly handler_digest: string;
  readonly input_type_ref: string;
  readonly output_type_ref: string;
  readonly return_mode: HookReturnMode;
  readonly body_region_id: string;
  readonly assigned_context_value_id?: string;
};

/** An explicit typed Context transition between two nodes. */
export type BoundContextEdge = {
  readonly from_node: string;
  readonly to_node: string;
  readonly context_type_ref: string;
  readonly value_id: string;
};

/**
 * One authored Capability declaration.
 *
 * Declarations are held one per authored binding, never one per
 * `capability_ref`: the same capability declared as both a Tool and a plain
 * Capability is two distinct requirements and both reach the FrontendGraph.
 */
export type BoundCapabilityRequirement = {
  readonly capability_ref: string;
  readonly tool_schema_present: boolean;
  readonly requested_permission?: Permission;
};

/** The immutable bound tree for one authored Agent program. */
export type BoundProgram = {
  readonly program_id: string;
  readonly entrypoint: string;
  readonly input_type_ref: string;
  readonly input_contract?: InputContract;
  readonly output_type_ref: string;
  readonly has_default_context: boolean;
  readonly context_type_ref?: string;
  readonly parameters: readonly BoundParameter[];
  readonly body_region_id: string;
  readonly declarations: readonly BoundDeclaration[];
  readonly values: readonly BoundValue[];
  readonly regions: readonly BoundRegion[];
  readonly calls: readonly BoundCall[];
  readonly controls: readonly BoundControl[];
  readonly context_edges: readonly BoundContextEdge[];
  readonly hooks: readonly BoundHook[];
  /** [program_ref, artifact_digest, entrypoint, target_agent_identity_requirement] */
  readonly imported_programs: readonly (readonly [string, string, string, string])[];
  readonly capability_requirements: readonly BoundCapabilityRequirement[];
  readonly model_requirements: readonly string[];
  /**
   * Authored Agent Skills, held as the generated contract record: unlike a
   * Capability requirement, nothing about a skill declaration is derived at
   * capture time, so there is nothing for a bound wrapper to add.
   */
  readonly skill_requirements: readonly SkillRequirement[];
  /** [node_id, span, semantic_annotation] */
  readonly spans: readonly (readonly [string, Span, string])[];
};

export type BoundNode = BoundCall | BoundControl;
